### DISCUSSION TITLE
Lab 02.03 - Task 02 Group Statistics

### DISCUSSION THREAD

**Discussion by Moorea-AI** [rramosp/ai4eng.v1#142] [02 - PYTHON]
Hola profe Raul y profe Felipe,

Estoy tratando de resolver la tarea 2 del laboratorio de Pandas y estoy confundida. No sé donde tengo el error y he tratado cambiando varias cosas y no me funciona. Podrian ayudarme? Muchas gracias!

Este es el código: 

def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
    categoria = np.array([0,1,2])
    media = np.array([df["price"].mean()])
    maximo = np.array([df["price"].max()])
    minimo = np.array([df["price"].min()])        
    
    result = pd.DataFrame(np.r_[[media, maximo, minimo]].T, index=categoria, columns=["media", "maximo", "minimo"])
    result.index.name="categoria"
    return result

Y este es el error:
error: ERROR EXECUTING YOUR CODE
---------------------------------
Traceback (most recent call last):
  File "calling_script.py", line 24, in <module>
    r = eval("vrskoznt.{}(source_functions, source_variables, '[aluz.moreno@udea.edu.co](mailto:aluz.moreno@udea.edu.co)')".format(grader_function_name), namespace)
  File "<string>", line 1, in <module>
  File "/content/vrskoznt.py", line 46, in grader_02
    answer = st_get_stats(d)
  File "<string>", line 10, in get_stats
  File "/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py", line 678, in __init__
    typ=manager,
  File "/usr/local/lib/python3.7/dist-packages/pandas/core/internals/construction.py", line 324, in ndarray_to_mgr
    _check_values_indices_shape_match(values, index, columns)
  File "/usr/local/lib/python3.7/dist-packages/pandas/core/internals/construction.py", line 393, in _check_values_indices_shape_match
    raise ValueError(f"Shape of passed values is {passed}, indices imply {implied}")
ValueError: Shape of passed values is (1, 3), indices imply (3, 3)
 (set session.debug=True for tracebacks)

**Comment by Felipe-RA** [comment]
Saludos Moorea

Lo primero, recuerda utilizar correctamente los  tags para agregar código, hace más fácil leerlo y usarlo.

Por ejemplo:

\```
def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
    categoria = np.array([0,1,2])
    media = np.array([df["price"].mean()])
    maximo = np.array([df["price"].max()])
    minimo = np.array([df["price"].min()])
    result = pd.DataFrame(np.r_[[media, maximo, minimo]].T, index=categoria, columns=["media", "maximo", "minimo"])
    result.index.name="categoria"
    return result
\```

El resultado de esto es el siguiente:

```
def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
    categoria = np.array([0,1,2])
    media = np.array([df["price"].mean()])
    maximo = np.array([df["price"].max()])
    minimo = np.array([df["price"].min()])
    result = pd.DataFrame(np.r_[[media, maximo, minimo]].T, index=categoria, columns=["media", "maximo", "minimo"])
    result.index.name="categoria"
    return result
```

Son 3 comillas arriba  y 3 comillas abajo

## Ahora tu pregunta:

El error es:  `ValueError: Shape of passed values is (1, 3), indices imply (3, 3)`

Al correr la celda de prueba, vemos este pedazo de error que nos indica en dónde está ocurriendo:

```
<ipython-input-10-cca1a630b819> in get_stats(df)
      6     maximo = np.array([df["price"].max()])
      7     minimo = np.array([df["price"].min()])
----> 8     result = pd.DataFrame(np.r_[[media, maximo, minimo]].T, index=categoria, columns=["media", "maximo", "minimo"])
      9     result.index.name="categoria"
     10     return result
```

Nos esta diciendo que `pd.DataFrame()` esperaba unos valores con dimensiones (osea con `shape = (3,3)`) pero está recibiendo valores con dimensión `shape = (1,3)`

Esto se traduce a que necesita una matrix 3x3 (el índice indica que debe tener  3x3) pero le estas dando una 1x3.

¿Por qué sabe él que debe ser 3x3?

Lo sabe porque le estas diciendo que el índice debe ser `categoria`, y como hay 3 categorias ya sabe que deben tenerse sólo 3 filas, y sabe que deben ser 3 columnas porque le estas indicando que debe tener `columns=["media", "maximo", "minimo"]`

### ¿Qué te recomiendo para solucionarlo?

Fijemonos en estas tres líneas:

```
media = np.array([df["price"].mean()])
maximo = np.array([df["price"].max()])
minimo = np.array([df["price"].min()])
```

Aquí lo que estas haciendo es crear un `array numpy` para la columna `precio`  con la media, uno con el máximo, y otro con el minimo.

Pero no estás discriminando entre las categorías, tienes que sacarle media, máximo y mínimo para cada categoría.

¿Cómo puedes hacer esto?

Los `pandas dataframes` soportan operaciones booleanas `in place`, esto es:

`df['category']==0`

En lenguaje natural lo podemos leer como: 
> Digame cuáles filas tienen categoría 0

Y python nos lo entrega de esta manera: (estos son datos que generé con un create df cualquiera, por supuesto los tuyos seran distintos)

```
itemid
74892    False
94636    False
77294     True
87061    False
24573    False
14855    False
70381     True
80538    False
79280     True
24907    False
Name: category, dtype: bool
```

Tenemos un dataframe de Booleanos, con `False` para los que no son categoría 0, `True` para los que son categoría 0.

Ahora, te recomiendo esta lectura para que puedas encontrar cómo utilizar esto en tu solución: [df.loc docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html)

Una vez entiendas cómo utilizar estas dos funciones, lo que debes hacer es generalizarlo para todas las categorias.

Buena suerte!
